flight/PiOS/Common/pios_bmi160.c

```diff
@@ -494,166 +494,202 @@ bool PIOS_BMI160_IRQHandler(void)
-			break;
-		case PIOS_BMI160_TOP_90DEG:
-			accel_data.x = accel_y;
-			accel_data.y = accel_x;
-			accel_data.z = -accel_z;
-			gyro_data.x  = gyro_y;
-			gyro_data.y  = gyro_x;
-			gyro_data.z  = -gyro_z;
-			break;
-		case PIOS_BMI160_TOP_180DEG:
-			accel_data.x = -accel_x;
-			accel_data.y = accel_y;
-			accel_data.z = -accel_z;
-			gyro_data.x  = -gyro_x;
-			gyro_data.y  = gyro_y;
-			gyro_data.z  = -gyro_z;
-			break;
-		case PIOS_BMI160_TOP_270DEG:
-			accel_data.x = -accel_y;
-			accel_data.y = -accel_x;
-			accel_data.z = -accel_z;
-			gyro_data.x  = -gyro_y;
-			gyro_data.y  = -gyro_x;
-			gyro_data.z  = -gyro_z;
-			break;
-		case PIOS_BMI160_BOTTOM_0DEG:
-			accel_data.x = accel_x;
-			accel_data.y = accel_y;
-			accel_data.z = accel_z;
-			gyro_data.x  = gyro_x;
-			gyro_data.y  = gyro_y;
-			gyro_data.z  = gyro_z;
-			break;
-		case PIOS_BMI160_BOTTOM_90DEG:
-			accel_data.x = -accel_y;
-			accel_data.y = accel_x;
-			accel_data.z = accel_z;
-			gyro_data.x  = -gyro_y;
-			gyro_data.y  = gyro_x;
-			gyro_data.z  = gyro_z;
-			break;
-		case PIOS_BMI160_BOTTOM_180DEG:
-			accel_data.x = -accel_x;
-			accel_data.y = -accel_y;
-			accel_data.z = accel_z;
-			gyro_data.x  = -gyro_x;
-			gyro_data.y  = -gyro_y;
-			gyro_data.z  = gyro_z;
-			break;
-		case PIOS_BMI160_BOTTOM_270DEG:
-			accel_data.x = accel_y;
-			accel_data.y = -accel_x;
-			accel_data.z = accel_z;
-			gyro_data.x  = gyro_y;
-			gyro_data.y  = -gyro_x;
-			gyro_data.z  = gyro_z;
-			break;
-		}
+	dev->temp_interleave_cnt += 1;
 
-		// Apply sensor scaling
-		accel_data.x *= dev->accel_scale;
-		accel_data.y *= dev->accel_scale;
-		accel_data.z *= dev->accel_scale;
-
-		gyro_data.x *= dev->gyro_scale;
-		gyro_data.y *= dev->gyro_scale;
-		gyro_data.z *= dev->gyro_scale;
-
-		// Get the temperature
-		// NOTE: We do this down here so the chip-select has some time to go low. Strange things happen
-		// When this is done right after readin the accels / gyros
-		if (temp_interleave_cnt % dev->cfg->temperature_interleaving == 0){
-			if (PIOS_BMI160_ClaimBus() != 0)
-				continue;
-
-			uint8_t bmi160_tx_buf[BUFFER_SIZE] = {BMI160_REG_TEMPERATURE_0 | 0x80, 0, 0, 0, 0, 0,
-					0, 0, 0, 0, 0, 0, 0};
-			if (PIOS_SPI_TransferBlock(dev->spi_id, bmi160_tx_buf, bmi160_rec_buf, 3) < 0) {
-				PIOS_BMI160_ReleaseBus();
-				continue;
-			}
+	return 0;
+}
 
-			PIOS_BMI160_ReleaseBus();
-			temperature =  23.f + (int16_t)(bmi160_rec_buf[2] << 8 | bmi160_rec_buf[1]) / 512.f;
-		}
+static bool PIOS_BMI160_callback_gyro(void *ctx, void *output,
+		int ms_to_wait, int *next_call)
+{
+	struct bmi160_dev *dev = (struct bmi160_dev *)ctx;
+
+	PIOS_Assert(dev);
+	PIOS_Assert(output);
 
-		accel_data.temperature = temperature;
-		gyro_data.temperature = temperature;
+	*next_call = 0;		/* Baby you know you can call me anytime */
 
+	if (PIOS_Semaphore_Take(dev->data_ready_sema, ms_to_wait) != true) {
+		return false;
+	}
+
+	if (PIOS_BMI160_parse_data(dev) < -1) {
```

Comment:
Maybe put a comment here saying something like "If temp readout fails (-1), we use the previous successful reading".